posix: pthread: leverage zephyr's synchronization and threading api

cfriedt opened the pull request:
**Note: USE_NEGATIVE_ERRNO compliance failures are false positives (pthread functions return +ve errno values)**

7 commits. Please see each commit message for details.

* posix: pthread: reimplement pthread_barrier using zephyr objects
* posix: mutex: use k_mutex instead of posix_mutex
* posix: cond: use k_condvar instead of posix_cond
* posix: pthread: rework pthreads to use zephyr api
* posix: pthread: add option for pthread_create() barrier
* tests: posix: improvements to pthread_pressure test
* posix: pthread: mitigate pthread include order sensitivity 

Fixes #58856
Fixes #56163
Fixes #57232

This significantly improves the reliability of pthreads and eliminates a race condition. Additionally, with this pthreads benefit from improvements to Zephyr's synchronization primitives and threading API going forward.

Commit:
posix: mutex: use k_mutex instead of posix_mutex

The internal representation of `pthread_mutex_t`,
`struct posix_mutex`, is basically a clone of `struct k_mutex`
but without the benefit of being able to use all of the
existing `k_mutex_*()` suite of functions.

The first step in the right direction was switching
the external representation of `pthread_mutex_t` to a simple
`int`. Let's take the next step in the right direction, which
is getting rid of `struct posix_mutex`.

The only significant difference between `struct k_mutex` and
`struct posix_mutex` is that the latter needs a `type` field.

Since there were a fixed number of `struct posix_mutex`, we
can just externalize the `type` field and reuse
`struct k_mutex` as-is.

For now, let's keep this change as a simple type
substitution. Eventually, we should be able to fully switch
to Zephyr API internally.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

Commit:
posix: cond: use k_condvar instead of posix_cond

The internal representation of `pthread_cond_t`,
`struct posix_cond`, is an identical clone of `struct k_condvar`
but without the benefit of being able to use all of the
existing `k_condvar_*()` suite of functions.

The first step in the right direction was switching
the external representation of `pthread_cond_t` to a simple
`int`. Let's take the next step in the right direction, which
is getting rid of `struct posix_cond`.

For now, let's keep this change as a simple type
substitution. Eventually, we should be able to fully switch
to Zephyr API internally.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

Commit:
posix: pthread: mitigate include order sensitivity

Previously, the `posix_internal.h` header needed to be exposed
to the application because we had non-trivial details for
most posix types (pthread, mutex, cond, ...). Since most of
those have been simplified to a typedef'ed integer, we
no longer need to expose that header to the applicaiton.

Additionally, it means that we can adopt normalized
header order in posix.

Additionally, keep more implementation details hidden
and prefer the static keyword on internal symbols where
possible.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

Commit:
posix: pthread: reimplement pthread_barrier using zephyr objects

Previously pthread_barrier_t was implemented in terms of wait
queues and internal scheduler functions.

This introduced some obstacles and inconsistency. In order
to be more consistent, rely only on Zephyr's public API and
reuse as many concepts as possible.

Deprecate `PTHREAD_BARRIER_DEFINE()` since it's non-standard.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

Commit:
posix: pthread: rework pthreads to use zephyr api

Previously, pthreads suffered from some race conditions.
This was almost inevitable given that it was maintained in
parallel to Zephyr's threading and synchronization API.

The unfortunate side-effect of with that is that it did not
receive the reliability and other improvements that
`k_thread`s did.

Here, we perform a significant update of pthread code so
that it depends directly on public Zephyr API. With that,
we reuse as many concepts as possible and pthreads benefits for
free from any improvement made to Zephyr's threading and
synchronization APIs.

Included with this change, we
* implement state with `ready_q`, `run_q`, and `done_q`
* use `pthread_barrier_wait()` to sync `pthread_create()`
* synchronize internal state with a spinlock

These pthreads are considerably more reliable than
before.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

Commit:
tests: posix: improvements to pthread_pressure test

* allow `qemu_cortex_a53`
* disallow `qemu_leon3
* remove `TEST_DURATION_S` range
* additional report formatting

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

cfriedt commented:
<img width="427" alt="Screenshot 2023-06-06 at 9 54 37 PM" src="https://github.com/zephyrproject-rtos/zephyr/assets/983494/967b70c0-fb82-43e2-ac42-fbc306f551e7">
👀

cfriedt commented:
* removed a bit of cross talk between commits 4 and 5

jgl-meta reviewed (dismissed):
Small comments. But everything looks good to me.

keith-packard commented on lib/posix/pthread_barrier.c:
```diff
@@ -4,31 +4,178 @@
+BUILD_ASSERT(CONFIG_MAX_PTHREAD_BARRIER_COUNT < PTHREAD_OBJ_MASK_INIT,
+	     "CONFIG_MAX_PTHREAD_BARRIER_COUNT is too high");
+
+static inline size_t posix_barrier_to_offset(struct posix_barrier *bar)
```
should use `ptrdiff_t` instead of `size_t`

keith-packard commented on lib/posix/pthread_barrier.c:
```diff
@@ -4,31 +4,178 @@
+	return mark_pthread_obj_uninitialized(barrier);
+}
+
+struct posix_barrier *get_posix_barrier(pthread_barrier_t b)
```
Should this do any range checking of the value to make sure it isn't completely wild?

cfriedt commented on lib/posix/pthread_barrier.c:
```diff
@@ -4,31 +4,178 @@
+	bit = posix_barrier_to_offset(bar);
+	err = sys_bitarray_free(&posix_barrier_bitarray, 1, bit);
+	__ASSERT_NO_MSG(err == 0);
+	*b = -1;
```
Meh - to be honest, this line isn't even necessary. Twister hasn't thrown any warnings, so if a cast was really necessary, we probably would have seen a warning / error.

cfriedt replied:
I removed it

keith-packard reviewed (changes requested):
This seems like a really nice clean up and simplification. I found a couple of minor type issues, but overall the design looks good.

cfriedt commented on lib/posix/pthread_barrier.c:
```diff
@@ -4,31 +4,178 @@
+
+	bar = get_posix_barrier(bb);
+	if (bar == NULL) {
+		return EINVAL;
```
Yes, exactly. It's the same with all of the `pthread` API.
https://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_barrier_destroy.html

cfriedt commented:
* use `false` instead of `0` in `bool` assignment
* remove unnecessary `*b = -1`
* change variable names `barrier -> b` to be more consistent

keith-packard commented on lib/posix/pthread_mutex.c:
```diff
@@ -112,18 +116,20 @@ static int acquire_mutex(pthread_mutex_t *mu, k_timeout_t timeout)
 		return EINVAL;
 	}
 
+	bit = posix_mutex_to_offset(m);
```
seems like this name is inaccurate; it's not the bit number, it's the index into the static array (yes, they're the same value, but used in different contexts, I think?)

cfriedt replied:
It's the bit in the bit array.

keith-packard commented on lib/posix/pthread_mutex.c:
```diff
@@ -112,18 +116,20 @@ static int acquire_mutex(pthread_mutex_t *mu, k_timeout_t timeout)
-		if (m->type == PTHREAD_MUTEX_RECURSIVE &&
+		type = posix_mutex_type[bit];
+		if (type == PTHREAD_MUTEX_RECURSIVE &&
 		    m->lock_count < MUTEX_MAX_REC_LOCK) {
```
Need to return EAGAIN if the lock count hits the limit.

keith-packard commented on lib/posix/pthread_cond.c:
```diff
@@ -218,3 +217,17 @@ int pthread_cond_destroy(pthread_cond_t *cvar)
+	int err;
+	size_t i;
+
+	for (i = 0; i < CONFIG_MAX_PTHREAD_BARRIER_COUNT; ++i) {
```
Should this use `CONFIG_MAX_PTHREAD_COND_COUNT`?

cfriedt commented on lib/posix/pthread_mutex.c:
```diff
@@ -327,3 +339,17 @@ int pthread_mutexattr_settype(pthread_mutexattr_t *_attr, int type)
+	int err;
+	size_t i;
+
+	for (i = 0; i < CONFIG_MAX_PTHREAD_BARRIER_COUNT; ++i) {
```
Indeed - `CONFIG_MAX_PTHREAD_MUTEX_COUNT`

I also fixed it in pthread_cond.c

keith-packard commented on lib/posix/posix_internal.h:
```diff
@@ -15,35 +15,25 @@
-	enum pthread_state state;
-	pthread_mutex_t state_lock;
-	pthread_cond_t state_cond;
+	/* Flags */
```
There are four bitfields here that are all <= 8 bits. I'd use bytes instead of bits as that's a lot faster and will consume no more memory.

keith-packard commented on lib/posix/pthread.c:
```diff
@@ -36,23 +56,68 @@ static const struct pthread_attr init_pthread_attrs = {
+	 * This differs from other posix object allocation strategies because they use
+	 * a bitarray to indicate whether an object has been allocated.
+	 */
+	actually_initialized = t->qid != POSIX_THREAD_READY_Q;
```
Seems like this is just doing a single load under the spin lock? Might be nice to allow for architectures for which this is an atomic operation (essentially all of them)?

cfriedt replied:
I'm not sure if we have an atomic load for `uint8_t`, and everywhere else that touches the queues uses the spinlock as well.

My main goal with this change was eliminating the remaining race condition.

There are likely lots of micro-optimizations to be made, but I would be happy to optimize later.

keith-packard replied:
yeah, good point. Get it working, then get it working fast.

keith-packard commented on lib/posix/pthread.c:
```diff
@@ -36,23 +56,68 @@ static const struct pthread_attr init_pthread_attrs = {
+	size_t bit;
+	struct posix_thread *t;
+
+	t = (struct posix_thread *)CONTAINER_OF(k_current_get(), struct posix_thread, thread);
```
Would it be good to have an assert that the current thread was actually a posix thread? There's no valid error return from this function, so it would be a fatal error.

cfriedt replied:
Oddly enough, we need to be able to call `pthread_self()` even with non-posix threads for comparison purposes.

keith-packard commented on lib/posix/pthread.c:
```diff
@@ -151,69 +286,85 @@ static void zephyr_thread_wrapper(void *arg1, void *arg2, void *arg3)
+			}
+			sys_dlist_remove(&t->q_node);
+			sys_dlist_append(&ready_q, &t->q_node);
+			t->qid = POSIX_THREAD_READY_Q;
```
t->qid seems to be covered by t->lock in some places, but I don't see that locked here. Seems like you probably need to just use q_lock everywhere and ensure that t->qid and the actual queue membership are synchronously updated under the same spinlock at the same time.

keith-packard commented on lib/posix/posix_internal.h:
```diff
@@ -15,35 +15,25 @@
-	PTHREAD_EXITED
-};
-
 struct posix_thread {
```
Given the use of multiple spinlocks to protect various parts of this struct, I'd love to see comments indicating which lock covers which fields.

keith-packard replied:
alternatively, just use a single spinlock for all thread data; given that it looks like most state transitions involve queue manipulation anyhow, I'm not seeing how the per-thread lock will gain much performance.

cfriedt replied:
There is only 1 spinlock `lock`

keith-packard replied:
Seems like qid and the queue itself are covered by the global q_lock? That was where I got confused, seeing code using different locks in ways that looked like they were supposed to be protecting the same data.

keith-packard commented on lib/posix/pthread.c:
```diff
@@ -151,69 +286,85 @@ static void zephyr_thread_wrapper(void *arg1, void *arg2, void *arg3)
+
+	key = k_spin_lock(&q_lock);
+	/* recycle old nodes if they are available */
+	do {
```
What is this outer do loop for? Just to declare safe_t? Is this a common pattern?

cfriedt replied:
Just an extra scope, because we don't need `safe_t` after this point.

keith-packard replied:
Sure, welcome to C, I guess. Could just use braces instead of a do though; I was a bit confused by the 'continue' inside the inner loop and wondering if you intended that to go back to the top of the 'do {' loop.

keith-packard commented on lib/posix/pthread.c:
```diff
@@ -151,69 +286,85 @@ static void zephyr_thread_wrapper(void *arg1, void *arg2, void *arg3)
+		struct posix_thread *safe_t;
+
+		SYS_DLIST_FOR_EACH_CONTAINER_SAFE(&done_q, t, safe_t, q_node) {
+			if (t->detachstate == PTHREAD_CREATE_JOINABLE) {
```
Could this transition be managed along with the manipulation of detachstate and/or moving objects into done_q? It seems a bit odd to have threads hanging in limbo until someone goes to create another thread.

keith-packard replied:
yeah, workqueue def seems like overkill. Was wondering if it could be done when the state transition is made though; you're just looking for threads in the done_q which are JOINABLE, so that should be detectable when either adding to done_q or switching to JOINABLE, it seems. Might be a future cleanup though; this solution looks correct.

keith-packard commented on lib/posix/pthread.c:
```diff
@@ -151,69 +286,85 @@ static void zephyr_thread_wrapper(void *arg1, void *arg2, void *arg3)
+	/* check if a thread is available to spawn */
+	t = NULL;
+	if (!sys_dlist_is_empty(&ready_q)) {
+		t = CONTAINER_OF(sys_dlist_get(&ready_q), struct posix_thread, q_node);
```
Should the thread be pulled out of the ready_q right here? Seems like there's a possible race with two pthread_create calls both ending up with the same thread?

cfriedt replied:
I don't think that's possible - in order to manipulate the which queue a thread is in (i.e. `q_node` within each `struct posix_thread`) `q_lock` needs to be held.

It's clearly held here for the entirety of the queue manipulation.

keith-packard replied:
q_lock is held here, I'm wondering why you're setting qid to `POSIX_THREAD_NONE_Q` without also removing `t` from `ready_q`. That seems like a bug.

keith-packard replied:
Especially when you've got a failure case down below that re-inserts it into `ready_q` and resets qid to `POSIX_THREAD_READY_Q`.

keith-packard replied:
Thanks, that answers my concern.

keith-packard commented on lib/posix/pthread.c:
```diff
@@ -151,69 +286,85 @@ static void zephyr_thread_wrapper(void *arg1, void *arg2, void *arg3)
+	err = pthread_barrier_init(&barrier, NULL, 2);
+	if (err != 0) {
+		key = k_spin_lock(&q_lock);
+		sys_dlist_append(&ready_q, &t->q_node);
```
oh, looks like this bit does expect the thread to have been pulled from ready_q?

keith-packard commented on lib/posix/pthread.c:
```diff
@@ -151,69 +286,85 @@ static void zephyr_thread_wrapper(void *arg1, void *arg2, void *arg3)
-	pthread_mutex_lock(&thread->state_lock);
-	thread->state = attr->detachstate;
-	pthread_mutex_unlock(&thread->state_lock);
+	err = pthread_barrier_wait(&barrier);
```
Ah, it looks like you're using the barrier to avoid synchronization bugs in the underlying k_thread implementation. If so, it would be good to mention that here -- it seems to me that a "correct" k_threads implementation would not require the thread to have started before being able to manipulate it with further k_thread API calls from the parent thread?

cfriedt replied:
`k_thread` is fine (I think? although there may be lingering arch-specific issues).

The thing that we want to synchronize is that the thread should be started and joinable by the time that `pthread_create()` returns.

Technically, as soon as `k_thread_create()` returns here, it should be possible for `k_thread_join()` to be called. The `k_thread` implementation is not the problem. It's actually `pthread_create()` and `pthread_join()` which are problematic - they do not operate in kernel context, and so there is no guarantee if `pthread_create()` will return before `pthread_join()` is called.

The barrier helps to synchronize that event though.

keith-packard replied:
I'm confused here (not surprising, this code is the most complex thing I've reviewed in months)...

In order for `pthread_join()` to be callable before `pthread_create()` returns, it would need the thread ID. You've very carefully ensured that the thread id is *not* visible until all of the work in `pthread_create()` has finished. So, it seems to me that you've already guaranteed that `pthread_join()` cannot be called until after `pthread_create()` is done (not quite returned, that's the next statement after the thread id assignment...).

Are you concerned about `pthread_join()` being called with the thread id from the last time the same slot was used?

cfriedt replied:
It's not so much that `pthread_join()` can't be called until `pthread_create()` has returned, it's that the new thread itself needs to be in a joinable state before `pthread_create()` returns (and before `pthread_join()` can be called).

That basically reduces to the pthread needing to be in a joinable state before `pthread_join()` can be called.

For pthreads (at least with this fancy new implementation), that means
1. the thread cannot be in `ready_q`
2. it must have been created `PTHREAD_CREATE_JOINABLE` (of course)

I think that's it..

keith-packard replied:
so the thread needs to be out of `ready_q` before `pthread_create()` returns. Is there some reason that couldn't happen? Given the cost of the extra context switches, it seems worth a bit of investigation to see if we can't avoid this synchronization ...

cfriedt replied:
`k_thread_create()` is a scheduling point, and there is no guarantee that this or the other would be the next context swapped in. At least, that's my guess. Is that a semi-accurate take, @andyross ?

keith-packard replied:
Ok, I think I understand things well enough now. As above, I *think* you could move the `run_q` assignment to this function, but that would require a bunch more thinking and it purely a performance optimization.

keith-packard commented on lib/posix/pthread.c:
```diff
@@ -226,22 +377,27 @@ int pthread_create(pthread_t *newthread, const pthread_attr_t *_attr,
+	k_spin_unlock(&t->lock, key);
 
 	if (state == PTHREAD_CANCEL_ENABLE && cancel_pending) {
-		pthread_exit((void *)PTHREAD_CANCELED);
```
From my reading of the POSIX spec, `PTHREAD_CANCELED` should already be a `void *` value, so no cast should be needed here. Might do some checking to make sure this is true -- otherwise, you wouldn't be able to test for this value with a simple compare operation like `res == PTHREAD_CANCELED` as the POSIX examples show.

cfriedt replied:
That is true! I'll remove the case 👍

keith-packard commented on lib/posix/pthread.c:
```diff
@@ -254,31 +410,32 @@ int pthread_setcancelstate(int state, int *oldstate)
-	k_spin_unlock(&thread->cancel_lock, cancel_key);
-
+	key = k_spin_lock(&t->lock);
+	t->cancel_pending = true;
```
Should this check to see if `cancel_pending` is already `true`?

cfriedt replied:
in the previous implementation it was using the same logic. The only pace where `cancel_pending` is checked from what I've seen is `pthread_setcancelstate()`

keith-packard commented on lib/posix/pthread.c:
```diff
@@ -410,54 +564,52 @@ void pthread_exit(void *retval)
+	int qid;
+	int detachstate;
+	k_spinlock_key_t key;
+	struct posix_thread *t;
```
would be nice for the `struct posix_thread *` variable in various functions to use the same name, along with other variables.

cfriedt replied:
true. I'll make that change.

keith-packard commented on lib/posix/pthread.c:
```diff
@@ -410,54 +564,52 @@ void pthread_exit(void *retval)
-		ret = EINVAL;
-	} else {
-		ret = ESRCH;
+	if (detachstate == PTHREAD_CREATE_JOINABLE && qid != POSIX_THREAD_READY_Q) {
```
Hrm. Is there a possible race with fetching qid under the lock and testing it with the lock released?

cfriedt replied:
should be fixed now.

keith-packard reviewed (commented):
Really liking this series, the changes to the pthread_cond make it clear that the overall plan is an improvement. I've got some concerns about using spinlocks instead of a mutex as that means the code can't call blocking APIs with locks held, which leaves me asking whether there are possible races between fetching state and acting on it.

keith-packard commented on include/zephyr/posix/pthread.h:
```diff
@@ -7,14 +7,15 @@
 #ifndef ZEPHYR_INCLUDE_POSIX_PTHREAD_H_
 #define ZEPHYR_INCLUDE_POSIX_PTHREAD_H_
 
+#include "pthread_key.h"
```
Probably shouldn't use " path here -- this is a public header and we don't want to accidentally pull in an application -specific header?

cfriedt replied:
It's not _really_ a public header - and what I mean by that is that it's a not a standard posix header.

In any case, it's not being added, removed or modified in this PR.

cfriedt commented on lib/posix/pthread.c:
```diff
@@ -135,12 +200,79 @@ int pthread_attr_setstack(pthread_attr_t *_attr, void *stackaddr, size_t stacksi
+	t = to_posix_thread(pthread_self());
+
+	/* cross the barrier so that pthread_create() can continue */
+	err = pthread_barrier_wait(&barrier);
```
Yep

cfriedt commented on lib/posix/pthread.c:
```diff
@@ -135,12 +203,79 @@ int pthread_attr_setstack(pthread_attr_t *_attr, void *stackaddr, size_t stacksi
+	void *arg;
+	k_spinlock_key_t key;
+	struct posix_thread *t;
+	pthread_barrier_t barrier;
```
It's an integer and can be passed via register

keith-packard replied:
ah, I totally missed how it got its value. Everything is "just an int now", which is a great simplifier.

cfriedt commented on lib/posix/pthread.c:
```diff
@@ -135,12 +200,79 @@ int pthread_attr_setstack(pthread_attr_t *_attr, void *stackaddr, size_t stacksi
+	__ASSERT_NO_MSG(err == 0 || err == PTHREAD_BARRIER_SERIAL_THREAD);
+
+	/* add to run_q */
+	key = k_spin_lock(&t->lock);
```
I guess I could add it to the run queue first.

cfriedt replied:
Fixed!

cfriedt commented:
Wow - tough review questions @keith-packard, but at least on my desk, your code suggestions helped me double the performance on `qemu_riscv64_smp`. Thanks! 👍 

FWIW, I left a few comments in `tests/posix/pthread_pressure/prj.conf` for benchmarking purposes.

<img width="640" alt="Screenshot 2023-06-07 at 5 36 51 PM" src="https://github.com/zephyrproject-rtos/zephyr/assets/983494/774ab5a9-6195-4b8e-bd39-b979ca5040e6">

keith-packard commented:
> Wow - tough review questions @keith-packard, but at least on my desk, your code suggestions helped me double the performance on `qemu_riscv64_smp`. Thanks! +1

You did ask for review. Happy to ask questions based on only a vague understanding of the code and have at least one of them turn out to be useful.

> There is always room for optimization down the road as well.

Yup. From a correctness perspective, I think this series is in great shape. I'll go do another brief pass through the types to make sure they're in good shape; I think the big issues I found was mixing up offset_t and size_t. It seems to be endemic though, so maybe I just need to accept it?

keith-packard reviewed (dismissed):
I think the only open concerns I have are performance related, not correctness. Given that this works way better than the existing code, let's get 'er merged and consider future performance improvements separately.

evgeniy-paltsev commented on lib/posix/posix_internal.h:
```diff
@@ -95,24 +68,9 @@ static inline uint32_t mark_pthread_obj_uninitialized(uint32_t obj)
-
-/* get a previously initialized posix_key */
-pthread_key_obj *get_posix_key(pthread_key_t key);
+struct k_mutex *to_posix_mutex(pthread_mutex_t *mu);
```
Shouldn't we change the name here? Previously this function did conversion to posix mutex, but now it converts to regular Zephyr mutex.

Bumped into this reviewing later code - as due to `to_posix_mutex` function name code looks incorrect.

cfriedt replied:
I didn't feel the need to rename functions / introduce additional noise.

cfriedt replied:
@evgeniy-paltsev - I would rather not change the pattern with these here. There are similar functions throughout the posix code and they are internal.

evgeniy-paltsev commented on lib/posix/pthread_mutex.c:
```diff
@@ -90,60 +91,62 @@ struct posix_mutex *to_posix_mutex(pthread_mutex_t *mu)
+				break;
+			}
+			/* On most POSIX systems, this usually results in an infinite loop */
+			do {
+				k_yield();
+			} while (true);
```
Hmm, as per previous explanation I understand that we expect current posix thread to hang in this case.

Do we expect that it will also break other threads (as we hold `pthread_mutex_spinlock` spinlock and never unlock it)?

evgeniy-paltsev replied:
Probably we should unlock spinlock and call k_sleep(K_FOREVER) here.

cfriedt replied:
Sure

evgeniy-paltsev commented on lib/posix/pthread_mutex.c:
```diff
@@ -90,60 +91,62 @@ struct posix_mutex *to_posix_mutex(pthread_mutex_t *mu)
+				ret = EAGAIN;
+			}
+			break;
+		case PTHREAD_MUTEX_ERRORCHECK:
+				ret = EDEADLK;
+			break;
```
Indentation.

```suggestion
		case PTHREAD_MUTEX_ERRORCHECK:
			ret = EDEADLK;
			break;
```

cfriedt replied:
Fixed

evgeniy-paltsev commented on lib/posix/pthread_mutex.c:
```diff
@@ -165,8 +168,7 @@ int pthread_mutex_trylock(pthread_mutex_t *m)
 {
-	int32_t timeout = (int32_t)timespec_to_timeoutms(abstime);
-	return acquire_mutex(m, K_MSEC(timeout));
+	return acquire_mutex(m, K_MSEC((int32_t)timespec_to_timeoutms(abstime)));
```
Feel free to ignore - but it looks like there is no functional changes here and we just make code much less readable.

cfriedt replied:
It saves a register 🤷🏼‍♂️

cfriedt replied:
Fixed!

evgeniy-paltsev commented on tests/posix/pthread_pressure/Kconfig:
```diff
@@ -15,7 +15,6 @@ config TEST_NUM_CPUS
-	range 1 21600
 	default 5
 	help
 	   Duration for the test, in seconds. The range has a reblatively high
```
We drop the range, but still mention its values in help.

cfriedt replied:
Fixed!

evgeniy-paltsev reviewed (dismissed):
Just a brief look.

evgeniy-paltsev commented:
Hi @cfriedt ,

just wondering if we actually planning to add all this to upcoming 3.4 release (as there is 3.4 milestone)?

cfriedt commented on lib/posix/pthread_barrier.c:
```diff
@@ -4,31 +4,177 @@
+	}
 
-	b->count++;
+	err = k_mutex_lock(&bar->mutex, K_FOREVER);
+	if (err < 0) {
+		return -err;
+	}
+	__ASSERT_NO_MSG(err == 0);
```
I think originally, I had it as a direct assert,  without the conditional. That would be my preference. I would have to go back to my laptop to see what the context is as to whether that is ok in this situation.

cfriedt replied:
The assert works

evgeniy-paltsev replied:
> The assert works

Probably I don't understand your comment. What do you mean by 'work'? Do you mean that it's not triggered? 

My concern here is that we **can't** possibly trigger it. `k_mutex_lock` as zephyr API return either 0 or negative error code. In case of error we return before the assert, so assert will always be executed with `err = 0`.

cfriedt commented on lib/posix/pthread.c:
```diff
@@ -36,23 +58,68 @@ static const struct pthread_attr init_pthread_attrs = {
+	actually_initialized = t->qid != POSIX_THREAD_READY_Q;
+	k_spin_unlock(&q_lock, key);
+
+	if (actually_initialized == 0) {
```
Fixed!

> Hi @cfriedt ,
> 
> just wondering if we actually planning to add all this to upcoming 3.4 release (as there is 3.4 milestone)?

I am planning to, yes

cfriedt commented:
> Hi @cfriedt ,
> 
> just wondering if we actually planning to add all this to upcoming 3.4 release (as there is 3.4 milestone)?

@evgeniy-paltsev - that was the plan, yes

cfriedt commented:
This is actually showing a 10-14% performance improvement fwict on `qemu_riscv64_smp`.

cfriedt commented:
@evgeniy-paltsev - please re-review when you have some time within the next day. Otherwise, @keith-packard, @jgl-meta, please re-ack when you have a moment.

keith-packard commented on lib/posix/pthread.c:
```diff
@@ -135,12 +202,79 @@ int pthread_attr_setstack(pthread_attr_t *_attr, void *stackaddr, size_t stacksi
+	pthread_barrier_t barrier;
+	void *(*fun_ptr)(void *arg);
+
+	arg = arg1;
```
ah, now I understand how I missed the assignment to barrier -- these three assignments could be done as initializers instead. Might be a nice clean up. If you like things that way (which I do).

cfriedt replied:
OK

keith-packard commented on lib/posix/pthread.c:
```diff
@@ -135,12 +202,79 @@ int pthread_attr_setstack(pthread_attr_t *_attr, void *stackaddr, size_t stacksi
+	fun_ptr = arg3;
+	t = to_posix_thread(pthread_self());
+
+	/* add to run_q */
```
reading the patch again, I *think* it would be better to add the thread to `run_q` in `pthread_create()` instead as that avoids a couple of possible races? I don't see any downside to it, and I think we could (eventually) remove the barrier that way. And that should recover the performance loss on uni-processor targets as we don't have a couple of extra context switches during thread startup. To be clear -- I'm suggesting that as a possible future optimization, not a requirement for merging the series as it stands today.

cfriedt replied:
done!

keith-packard commented on lib/posix/pthread.c:
```diff
@@ -151,69 +285,83 @@ static void zephyr_thread_wrapper(void *arg1, void *arg2, void *arg3)
-			thread->state = PTHREAD_JOINABLE;
-			break;
+	/* copy the attribute in to the stack */
+	(void)memcpy(&attr, _attr, sizeof(attr));
```
What's the reason for this copy? I don't see any uses of this value outside of the context of this function?

cfriedt replied:
I removed it - it was just paranoia

keith-packard commented on lib/posix/pthread.c:
```diff
@@ -254,31 +407,32 @@ int pthread_setcancelstate(int state, int *oldstate)
 
-		k_thread_abort(&thread->thread);
+	if (cancel_state == PTHREAD_CANCEL_ENABLE) {
+		posix_thread_finalize(t, retval);
```
This appears to assume that `pthread_cancel()` is only called on `pthread_self()`, which I don't believe is true. I think that means `posix_thread_finalize()` shouldn't call `k_thread_abort()` and that `pthread_exit()` should do that instead, after calling `posix_thread_finalize()`.

cfriedt replied:
Will check it out.

cfriedt replied:
Yes - good catch - fixed!

keith-packard commented on lib/posix/pthread.c:
```diff
@@ -254,31 +407,32 @@ int pthread_setcancelstate(int state, int *oldstate)
+	t->cancel_pending = true;
+	cancel_state = t->cancel_state;
+	detachstate = t->detachstate;
 	if (cancel_state == PTHREAD_CANCEL_ENABLE) {
```
I think the whole retval management is incorrect here -- if the thread is canceled, then the retval seen through `pthread_join()` should be `PTHREAD_CANCELED`. So, I think this chunk of code can simply be deleted and the value `PTHREAD_CANCELED` be passed unconditionally to `posix_thread_finalize()`. That code should make sure the thread is not already terminated before assigning the new retval to `t->retval` so that calls to `pthread_cancel()` don't override a simultaneous call to `pthread_exit()`.

keith-packard commented on lib/posix/pthread.c:
```diff
@@ -375,30 +530,26 @@ int pthread_once(pthread_once_t *once, void (*init_func)(void))
-	} else {
-		self->state = PTHREAD_TERMINATED;
-	}
+	/* Make a thread as cancelable before exiting */
```
Ok, this looks right -- if `pthread_cancel()` is called on this thread while the destructors are being processed, then we'll never get to the `posix_thread_finalize()` call below. Do we want to mark the thread as not cancelable before calling `posix_thread_finalize()` so that process doesn't get interrupted mid-way?

cfriedt replied:
Sounds reasonable

cfriedt replied:
I put the destructor calls into `posix_thread_finalize()` and the passed the thread to terminate to that function instead of `pthread_self()`.

keith-packard reviewed (commented):
I think pthread_cancel is busted in the current patch as it assumes threads will only call it on themselves.

cfriedt commented:
I was hacking a bit this morning on the possibility of removing the barrier but the stability is still unfortunately not there yet.

It's possible that it points at underlying memory, caching, qemu things, but it's also something that we can fix later.

I'll make a few changes suggested w.r.t. cancellation / coding style, but I think I'm likely not going to try any additional optimizations at this point.

cfriedt commented:
I think I've addressed most comments. Tests are passing. We'll put off removing the barrier for another day.

ycsin reviewed (dismissed):
LGTM, we can probably get this through first since it is an obvious improvement over the existing code.

keith-packard reviewed (dismissed):
Looks good.

jgl-meta commented:
@nashif I think you will have to merge this one.

andyross commented on lib/posix/pthread_mutex.c:
```diff
@@ -90,60 +91,62 @@ struct posix_mutex *to_posix_mutex(pthread_mutex_t *mu)
+				ret = EBUSY;
+				break;
+			}
+			/* On most POSIX systems, this usually results in an infinite loop */
```
Just to jump in on the resolved conversation: it does indeed seem like POSIX **specifies** a deadlock as the correct behavior here.  So I guess this code is needed...

FWIW: a somewhat smaller way to make sure _current never runs again is to do `k_thread_suspend(k_current_get());`.  No need to run that in a loop as it doesn't take a timeout and as an async API doesn't have a return value.

andyross reviewed (dismissed):
A handful of notes, nothing major (though that pthread_self() weirdness might want some documentation, it has the feel of a footgun).

cfriedt commented on lib/posix/posix_internal.h:
```diff
@@ -15,35 +15,23 @@
-	enum pthread_state state;
-	pthread_mutex_t state_lock;
-	pthread_cond_t state_cond;
+	/* Flags */
```
I blame @keith-packard - hehe... kidding.

cfriedt replied:
Fixed!

cfriedt commented on lib/posix/pthread.c:
```diff
@@ -151,73 +282,83 @@ static void zephyr_thread_wrapper(void *arg1, void *arg2, void *arg3)
+	k_thread_create(&t->thread, attr->stack, attr->stacksize, zephyr_thread_wrapper,
+			(void *)arg, UINT_TO_POINTER(barrier), threadroutine,
+			posix_to_zephyr_priority(attr->priority, attr->schedpolicy),
+			(~K_ESSENTIAL & attr->flags), K_MSEC(attr->delayedstart));
```
I'm not sure - that's how it was done with the original implementation, so i didn't want to veer off too far.

cfriedt replied:
Oh wait - I guess the flags may be relevant to Zephyr but specifically `K_ESSENTIAL` wouldn't work because then we wouldn't be able to abort the thread (I guess that would trigger a panic (see [z_thread_abort()](https://github.com/zephyrproject-rtos/zephyr/blob/028e0a3c8d75fc5e2027f54dbb5a55aadbb1c490/kernel/sched.c#L1736))

cfriedt replied:
I added a check in `pthread_attr_is_valid()`

cfriedt commented on lib/posix/pthread.c:
```diff
@@ -375,131 +513,107 @@ int pthread_once(pthread_once_t *once, void (*init_func)(void))
+	/* Make a thread as cancelable before exiting */
+	key = k_spin_lock(&q_lock);
+	self->cancel_state = PTHREAD_CANCEL_ENABLE;
+	k_spin_unlock(&q_lock, key);
```
Agreed. I just wanted to make it a bit less confusing for the average reader. Would be happy to make that optimization later.

keith-packard replied:
in particular, until the remaining instability is fixed, introducing weaker synchronization mechanisms like atomics seems like something to push out for now.

cfriedt commented:
I can do a quick respin addressing @andyross's comments if people are on-board for re-acking.

@jgl-meta if you would like to get this into v3.4.0, you may need to help to convince @nashif and @fabiobaltieri .

I'm easy either way.

There are SMP races with pthreads in `main` though, which this addresses.

If we would prefer holding off merging, I think this is at a good point to potentially investigate if the barrier is necessary and why some SMP platforms have issues when it is removed. Again, things get very hand-wavy when qemu and SMP are involved, but who knows - maybe these are very subtle wrinkles things that can be ironed-out with the odd DMB in arch code.

cfriedt commented on lib/posix/pthread.c:
```diff
@@ -375,131 +513,107 @@ int pthread_once(pthread_once_t *once, void (*init_func)(void))
-		ret = ESRCH;
-		break;
-	default:
+	key = k_spin_lock(&q_lock);
```
Yep - that could definitely be renamed. The original pthread implementation had multiple locks, and it took some convincing for me to reduce it to 1, care of Josh Lilly and Keith Packard.

keith-packard commented:
> I just need to more eyes and more reviews on this, we still have time to get things fixed up and looked at.

Eyes on the code sure, but do you think we need more experience with running code both in qemu and on real hardware? Or more tests exploring other aspects of the API?

cfriedt commented:
> > @jgl-meta if you would like to get this into v3.4.0, you may need to help to convince @nashif and @fabiobaltieri .
> 
> I just need to more eyes and more reviews on this, we still have time to get things fixed up and looked at.

OK - I'll respin again to address Andy's comments.

cfriedt commented on lib/posix/pthread.c:
```diff
@@ -375,131 +513,107 @@ int pthread_once(pthread_once_t *once, void (*init_func)(void))
+		return ret;
+	}
+
+	return 0;
```
Fixed!

Commit:
posix: pthread: add option for pthread_create() barrier

To enable testing, introduce `CONFIG_PTHREAD_CREATE_BARRIER`.

Some observations were made that running several Qemu SMP targets
concurrently could lead to synchronization problems. On such
targets, it was found that the synchronization issues were
mitigated by introducing a `pthread_barrier_t` shared between
`pthread_create()` and the spawned thread.

It is suggested to enable the option when running many
SMP tests concurrently in several parallel Qemu processes,
e.g. with `twister`.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

cfriedt commented:
* added implementation and Kconfig option `CONFIG_PTHREAD_CREATE_BARRIER`

cfriedt commented:
* corrected a comment

cfriedt commented:
![chefs-kiss-small-er](https://github.com/zephyrproject-rtos/zephyr/assets/983494/6baf4492-ab2a-48bd-a89b-de5fff2118c4)
```
INFO    - Building initial testsuite list...
INFO    - Writing JSON report /home/cfriedt/zephyrproject/zephyr/twister-out/testplan.json
INFO    - JOBS: 32
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    - Total complete:  232/ 232  100%  skipped:   84, failed:    0, error:    0
INFO    - 29 test scenarios (232 test instances) selected, 84 configurations skipped (84 by static filter, 0 at runtime).
INFO    - 148 of 232 test configurations passed (100.00%), 0 failed, 0 errored, 84 skipped with 0 warnings in 97.84 seconds
INFO    - In total 2488 test cases were executed, 2184 skipped on 8 out of total 573 platforms (1.40%)
INFO    - 140 test configurations executed on platforms, 8 test configurations were only built.
INFO    - Saving reports...
INFO    - Writing JSON report /home/cfriedt/zephyrproject/zephyr/twister-out/twister.json
INFO    - Writing xunit report /home/cfriedt/zephyrproject/zephyr/twister-out/twister.xml...
INFO    - Writing xunit report /home/cfriedt/zephyrproject/zephyr/twister-out/twister_report.xml...
INFO    - Run completed
```
<img width="639" alt="Screenshot 2023-06-08 at 8 45 38 PM" src="https://github.com/zephyrproject-rtos/zephyr/assets/983494/20826361-8d91-408f-ae42-d432877a084e">

cfriedt commented:
There was some discussion on Discord - in any case, to me it really seemed to be less of a Zephyr problem and more of a problem from running many parallel SMP Qemu processes concurrently while having few assertions enabled, host scheduling noise, etc.

That effect vary based on host CPU. It's hard to say.

The default configuration is now 
* `CONFIG_PTHREAD_CREATE_BARRIER=n`
* `CONFIG_TEST_EXTRA_ASSERTIONS=y` (in `tests/posix/pthread_pressure/`)

I would say, in combination with the above, as well as
* not gonna lie, amazing job done increasing pthread stability with the help of many reviewers 🙏

This is looking really good.

Thanks everyone 👍

cfriedt commented:
This minor micro-optimization creates a 50% performance gain on `qemu_riscv64_smp`.
* skip validating thread pointer in `zephyr_thread_wrapper()` (only called from `pthread_create()`)

Sorry to add yet another rev 😅 - probably can't do a lot of harm at this point though.

```diff
diff --git a/lib/posix/pthread.c b/lib/posix/pthread.c
index c55e824d35..30044420f1 100644
--- a/lib/posix/pthread.c
+++ b/lib/posix/pthread.c
@@ -268,6 +268,7 @@ static void zephyr_thread_wrapper(void *arg1, void *arg2, void *arg3)
        int err;
        int barrier;
        void *(*fun_ptr)(void *arg) = arg2;
+       struct posix_thread *t = CONTAINER_OF(k_current_get(), struct posix_thread, thread);
 
        if (IS_ENABLED(CONFIG_PTHREAD_CREATE_BARRIER)) {
                /* cross the barrier so that pthread_create() can continue */
@@ -276,7 +277,7 @@ static void zephyr_thread_wrapper(void *arg1, void *arg2, void *arg3)
                __ASSERT_NO_MSG(err == 0 || err == PTHREAD_BARRIER_SERIAL_THREAD);
        }
 
-       pthread_exit(fun_ptr(arg1));
+       posix_thread_finalize(t, fun_ptr(arg1));
 
        CODE_UNREACHABLE;
 }
```

andyross reviewed (approved):
Refresh +1
